File: leetcode/982.按位与为零的三元组.cpp
#include<immintrin.h>

__attribute__((no_sanitize_address,no_sanitize_memory))
__attribute__((target("avx2")))
void fwt(int* seq, int n) {
    const auto end = seq + n;
    for (int i = 1;i < 8 && i < n;i *= 2)
        for (auto j = seq;j != end;j += i)
            for (const auto k = j + i;j < k;++j)
                *j += j[i];
    const int m = n / 8;
    const auto l = reinterpret_cast<__m256i*>(seq);
    const auto r = reinterpret_cast<__m256i*>(end);
    for (int i = 1;i < m;i *= 2)
        for (auto j = l;j != r;j += i)
            for (const auto k = j + i;j < k;++j)
                *j = _mm256_add_epi32(*j, j[i]);
}

class Solution {
public:
    int countTriplets(const vector<int>& nums) {
        const int m = *max_element(nums.begin(), nums.end());
        const int l = m > 0 ? 1 << (32 - __builtin_clz(m)) : 1;
        alignas(32) int seq[l];
        fill_n(seq, l, 0);
        for (int e : nums)
            ++seq[e];
        fwt(seq, l);
        int ans = 0;
        for (int i = 0;i <= m;++i)
            ans += (1 - 2 * __builtin_parity(i)) * seq[i] * seq[i] * seq[i];
        return ans;
    }
};

File: leetcode/982.按位与为零的三元组.py
"""
枚举所有nums[i] & nums[j] 的个数，采用hash表进行统计
然后遍历k，如果 & 为零，则直接统计个数

"""
class Solution:
    def countTriplets1(self, nums: List[int]) -> int:
        n = len(nums)
        ans = 0
        cnt = Counter(x & y for x in nums for y in nums)
        for x in nums:
            for ky, val in cnt.items():
                if x & ky == 0:
                    ans += val
        return ans
    def countTriplets(self, nums: List[int]) -> int:
        cnt = [0] * (1 << 16)
        for x in nums:
            for y in nums:
                cnt[x & y] += 1
        ans = 0
        for m in nums:
            m ^= 0xffff
            ans += cnt[m]
            s = m
            while s:
                s = (s - 1) & m
                ans += cnt[s]
        return ans